### PR TITLE
[spec] Keep DSPARK draft tokens within the target vocabulary

### PR DESCRIPTION
**Pull request by klshuster**
## Motivation

DSPARK speculative decoding can emit and consume out-of-range token ids, which surface as runtime errors or device-side asserts (`probability tensor contains either inf, nan or element < 0`, scatter/gather index asserts) that bring down the whole engine — the failing rank's CUDA context dies and peer TP ranks abort their communicators. Two variants of the same class:

1. **Padded draft heads**: when the draft head is padded wider than the target's true vocabulary, draft sampling can propose pad-range ids, which then flow through verify into token gathers.
2. **Padding-independent verify bookkeeping**: chain-verify buffers are allocated with `torch.empty` and rows with no verify candidates (padded CUDA-graph rows, `verify_lens == 0`) are never written by the accept kernel — their uninitialized slots still feed the bonus-token gather; `correct_len` can go to `-1` unclamped; and rejected drafts are propagated into `out_tokens`. This variant fires even with an unpadded draft head and is load-dependent, since the buffers grow (fresh uninitialized memory) when the batch reaches a new size tier.

Both variants are covered by tests included in this PR (see Accuracy Tests).

## Modifications

- Cap draft sampling to the target's true vocabulary at all three entry points (eager block sampling, folded sampler, fused-greedy path — the last via a `vocab_limit` that slices base logits and the step-bias rows together).
- Crop the verify-side proposal distribution `q` (`corrected_logits`) to the same capped vocabulary that was sampled. With a padded head this corrects q's normalization (the softmax previously included pad rows that were never sampleable); with an unpadded head both the cap and the crop are no-op widths.
- Clamp `correct_len >= 0` for `verify_lens == 0` rows.
- Zero-initialize the chain-verify buffers (`predicts`, `accept_index`, `accept_token_num`) so first-use padded rows read a benign index instead of uninitialized memory.
- Zero-fill rejected and tail slots in `out_tokens` instead of propagating rejected drafts.

## Accuracy Tests

- GPU kernel parity suite extended (`test/registered/spec/dspark/test_dspark_kernel_parity.py`): two new subtests fail on the base commit on real behavior — `verify_lens == 0` yields `correct_len == -1` unclamped, and rejected/tail slots stay populated in `out_tokens` — and pass with this PR. The existing 20 parity subtests verify the accept path is unchanged for valid rows.
- CPU unit tests (`test/registered/unit/spec/test_dspark_vocab_bounds.py`): `test_chain_verify_buffers_start_zeroed` asserts the buffers start zeroed (fails on the base commit's `torch.empty` allocation in practice; `torch.empty` contents are unspecified, so that red is not guaranteed). `test_eager_draft_block_caps_proposals_and_q` shows a padded head whose pad rows carry the largest logits cannot be sampled past the true vocabulary, and that q is cropped to the same width (the capping parameter is new in this PR, so this test exercises the new path rather than reproducing the old failure).

## Speed Tests and Profiling

The cap is a logits slice (no-op width when the head is unpadded); buffer zero-init happens only on (re)allocation; the clamp and zero-fills are elementwise ops on tensors the verify path already touches.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (No user-facing surface changed.)
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). (Bounds/initialization guards; correctness covered by the kernel parity suite and unit tests above.)
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35659364476](https://github.com/sgl-project/sglang/actions/runs/35659364476)
Latest PR Test (Extra): :x: [Run #35659364340](https://github.com/sgl-project/sglang/actions/runs/35659364340)
Latest PR Test (AMD ROCm 10): :x: [Run #35659364517](https://github.com/sgl-project/sglang/actions/runs/35659364517)

### CODE REVIEW

File: python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py [hunk 1]
```diff
@@ -132,14 +138,17 @@ def __call__(self, hidden_states, input_ids):
         )
         if not self.folded_sampling and fused_greedy:
             draft_tokens = self.markov_head.sample_block_greedy_fused(
-                base_logits, first_prev_tokens=anchor
+                base_logits,
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py:141]: will this make https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/deepseek_v4_dspark.py#L517 break? shall we add this attribute there?

File: python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py [hunk 2]
```diff
@@ -199,7 +222,8 @@ def _resolve_folded_sampling(
     if getattr(model.markov_head, "supports_sharded_greedy", False):
         return False
     vocab = int(model.lm_head.org_vocab_size)
-    noise_bytes = max_bs * vocab * 4
+    capped_vocab = min(vocab, sample_vocab_size or vocab)
+    noise_bytes = max_bs * capped_vocab * 4
     logits_bytes = max_bs * gamma * vocab * _base_logits_dtype(model).itemsize
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py:227]: shall we change `vocab` here to `capped_vocab` too?
